### PR TITLE
More optimal examples of reduce usage

### PR DESCRIPTION
**Pull request by cytrowski**
### Description

I updated the examples where we calculate distribution of values in the array using reduce and how the groupBy is implemented. The previous examples were suboptimal because they unnecessarily created extra arrays on each iteration.

### Motivation

I often find devs using those code snippets in real-life projects (since redux is considered HARD to understand) and then later discovering performance issues with their code.

### Additional details

There is no extra context. I just believe the world will be better with those :)

### Related issues and pull requests

None - it's a spontaneous fix :)

### CODE REVIEW

**Comment by Josh-Cena** [comment]
We have this in the description:

> `reduce()` is a central concept in [functional programming](https://en.wikipedia.org/wiki/Functional_programming), where it's not possible to mutate any value, so in order to accumulate all values in an array, one must return a new accumulator value on every iteration. This convention propagates to JavaScript's `reduce()`: you should use [spreading](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) or other copying methods where possible to create new arrays and objects as the accumulator, rather than mutating the existing one. If you decided to mutate the accumulator instead of copying it, remember to still return the modified object in the callback, or the next iteration will receive undefined.

If you find yourself mutating your parameter, it means you are not practicing functional programming, and you'll find yourself better off (as in, writing both more performant and more readable code) by not using `reduce()` at all. Mutating anything in the callback is an anti-pattern, even in the name of performance.

**Comment by cytrowski** [comment]
@Josh-Cena with all due respect I totally disagree with this statement:

> reduce() is a central concept in [functional programming](https://en.wikipedia.org/wiki/Functional_programming), where it's not possible to mutate any value, so in order to accumulate all values in an array, one must return a new accumulator value on every iteration.

I agree that if you want to be FP-purist you can enforce those rules in your team. However it's JavaScript and In case of reduce it's totally fine to mutate the accumulator in the ways I proposed. There is no benefit of using spread operator in those cases. I consider those examples harmful. Also there is no sense of talking about FP-programming here when we don't have immutable data structures in JS, providing mutation-safe methods and updating the collections in the optimal way.

I'd appreciate if the examples on MDN were focused on practical usage of JavaScript. There is a lot of people learning to code this way and I believe this document deserves at least mentioning that it is suboptimal.

For those who look for immutable data structures and real FP style I recommend looking at: 

- https://immutable-js.com/
- https://ramdajs.com/

**Comment by Josh-Cena** [comment]
If you have the freedom of using `push()`, why would you write:

```js
const allbooks = friends.reduce(
  (accumulator, currentValue) => {
    accumulator.push(...currentValue.books);
    return accumulator;
  },
  ["Alphabet"],
);
```

Instead of this:

```js
const allbooks = ["Alphabet"];
for (const { books } of friends) {
  allbooks.push(...books);
}
```

The second one is much more readable and more performant. The one and only use case of `reduce` is when you don't want to mutate the accumulator, which makes a for loop useless.

FYI, the code was first changed in https://github.com/mdn/content/pull/20440. I totally agree with that change, and it's not constructive to revert someone else's work when we have reached agreement. I have yet to see a valid case where `reduce` with mutation is a genuine improvement over the equivalent code with `for...of`.

**Comment by cytrowski** [comment]
I believe this is a matter of code style. If the codebase you work with "prefers" array method chaining over for loops it's more feasible to use the optimal variant of reduce over the `for ... of ...` loop. In other words - I still think we should show better examples of reduce or maybe mention all the variants.

Also, because of the specific nature of the reduce accumulator, mutating it won't have any impact on the "external world". It's more of a temporary helper variable you return as a result of reduce call.

Also if we talk about "agreement" in the previous PR:

> This is more inline with how many popular linters will make users change their code, and is more "functional" in nature

I don't see those linters mentioned and I've never experienced my linter pinging me about this usage.

> Well, if you declared your input as readonly then TypeScript would catch that, but I'm reluctant to call it a linter

I agree with this :) 

I remember the previous examples and I was really surprised discovering those changes this week.

**Comment by Josh-Cena** [comment]
> Also, because of the specific nature of the reduce accumulator, mutating it won't have any impact on the "external world". It's more of a temporary helper variable you return as a result of reduce call.

This I agree; but therefore tolerating local mutations is against how `reduce()` works and its reason of existence. Had we been living in a mutating world, we would not even have the concept of `reduce`.

> If the codebase you work with "prefers" array method chaining over for loops it's more feasible to use the optimal variant of reduce over the `for ... of ...` loop

If performance is critical, you should not chain array methods at all—multiple intermediate arrays and running multiple callbacks is already ill-performing. If your array's size is reasonable (<10000), this optimization would probably be microscopic and the time is better invested elsewhere.

Again, we'd like MDN examples to demonstrate how an API _ought to_ be used, not just how it _is_ used. While I agree there are some performance gains, if you can tolerate mutations, you should always consider refactoring it to a `forEach` loop or a `for...of` loop. (And `forEach` also allows chaining.) `reduce` should _only_ be used to facilitate immutable programming.

**Comment by cytrowski** [comment]
> tolerating local mutations is against how reduce() works and its reason of existence

So what's the reason of existence for `reduce` then? You already told me using `for` is more readable. And now you are telling me that chaining of array methods is considered bad. 

I know that proper chaining implementation should allow streaming the array and break the iteration once the result is satisfying. Eg. `someNumbers.map(multiply).take(3)` should iterate only over first 3 elements. On the other hand I can always use `someNumbers.slice(3).map(multiply)` and be fine.

Chaining is very useful in scenarios where based on one source array we need to generate multiple different collections. Combined with proper memoization strategy it helps saving a lot of code boilerplate.

**Comment by Josh-Cena** [comment]
You can chain array methods, but beware that's already not performant—just that. And if you want to both chain methods and get the semantics of `for...of`, try `forEach`.

```js
const allbooks = ["Alphabet"];
friends.map(...).filter(...).forEach(({ books }) => allbooks.push(...books));
```

I'm not saying chaining is bad; I'm saying using `reduce` and doing mutations are contradictory in principle because that's not what `reduce` is designed for.

**Comment by cytrowski** [comment]
> we'd like MDN examples to demonstrate how an API ought to be used, not just how it is used

So you telling me that `reduce`, when it was introduced for the first time (Firefox 3, year 2008) was supposed to be used with what syntax? Spread for objects arrived quite recently (Chrome 60, 2017), and `Object.assign` was not there until the late 2014 (Firefox 34).

Using `forEach` from FP perspective is reserved for side effects once you finish all the "chained" calculations.

**Comment by cytrowski** [comment]
For posterity - I recommend this stackoverflow thread: https://stackoverflow.com/questions/63727586/is-mutating-accumulator-in-reduce-function-considered-bad-practice

TL;DR; Mutating values you don't own is bad. In case of reduce, with some exceptions, we own the accumulator. If function mutates the values it owns (no other function references them) - it's totally safe and more performant (both in terms of time and memory).

If you are looking for some "absolute rule" to make lazy decisions about your code then sure - go with: "never mutate" until the performance becomes really bad. My personal take here is that you shouldn't pick FP rules at random and JS is not good with neither FP nor OOP (if you use the Array.prototype API).

From what I learned during this discussion the better rule about `reduce` should be: "just don't use it if you chase readability or performance".

I strongly believe the way JS is built lets us pick the best tools for the job from both pure FP and more imperative style. You can write functions which do imperative, optimal work under the hood and expose FP-friendly API to compose more readable core of your app.

I'm sorry for digging so much here but I just don't buy the argumentation I received regarding this change. I know I'm not the core maintainer here and I appreciate the work you do here but I just don't agree with this stuff and I find the September change harmful for this doc.

I wont push this change further. Thanks for the discussion @Josh-Cena :)

**Comment by cytrowski** [comment]
Extra sidenote: I think I ❤️  this solution the most https://github.com/prateek3255/blog/issues/20#issuecomment-861398459
